files/common/usr/bin/illumos-cleanup

```diff
@@ -0,0 +1,185 @@
+}
+
+function destroy_dataset() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy "$dataset"; then
+			echo "Destroyed '$dataset'"
+		else
+			echo_err "Failed to destroy '$dataset'"
+			return 1
+		fi
+	fi
+}
+
+function destroy_dataset_recursive() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy -r "$dataset"; then
+			echo "Destroyed '$dataset' recursively"
+		else
+			echo_err "Failed to destroy '$dataset' recursively"
+			return 1
+		fi
+	fi
+}
+
+function promote_dataset() {
+	local dataset="$1"
+	local origin
+
+	#
+	# To promote a dataset all the way to the top, zfs promote may need
+	# to be called multiple times if the dataset is a nested clone.
+	# In practice though, since we delete old os versions, the clone
+	# depth shouldn't go deeper than 2 or 3 levels.
+	#
+	for _ in {1..5}; do
+		if ! origin=$(zfs get -H -o value origin "$dataset"); then
+			echo_err "Failed to get origin of '$dataset'"
+			return 1
+		fi
+		if [[ "$origin" == "-" ]]; then
+			echo "Promoted '$dataset'"
+			return 0
+		else
+			if ! zfs promote "$dataset"; then
+				echo_err "Failed to promote '$dataset'"
+				return 1
+			fi
+		fi
+	done
+
+	echo "Failed to promote '$dataset' to the top in 5 attempts"
+	echo "Dataset's origin is '$origin'"
+
+	return 1
+}
+
+#
+# Once we pass the no-rollback point after a migration, we want to
+# cleanup Illumos datasets that are not needed anymore.
+#
+function post_migration_cleanup() {
+	local var_delphix
+	local errors=false
+
+	if [[ ! -f "$POST_MIGRATION_CLEANUP" ]]; then
+		echo "Post-migration cleanup has already been performed."
+		return 0
+	fi
+
+	var_delphix=$(zfs list -H -o name /var/delphix) ||
+		die "Failed to get /var/delphix dataset"
+
+	#
+	# Clear the user-property generated by the Illumos installer
+	#
+	zfs inherit org.opensolaris.caiman:install rpool || errors=true
+
+	#
+	# Those datasets are cloned from Illumos datasets. We do not
+	# want any dependencies on Illumos datasets remaining after
+	# the migration.
+	#
+	promote_dataset rpool/illumos-home || errors=true
+	promote_dataset "$var_delphix" || errors=true
+
+	#
+	# We do not need dump block devices on Linux.
+	# We also do not currently support zvol swap devices on Linux.
+	#
+	destroy_dataset rpool/dump || errors=true
+	destroy_dataset rpool/swap || errors=true
+	destroy_dataset rpool/swap1 || errors=true
+
+	#
+	# Destroy temporary snapshots created for the duration of the
+	# migration.
+	#
+	destroy_dataset domain0/mds@migration || errors=true
+
+	#
+	# Some directories still have the old delphix:staff permissions.
+	# We want to convert those to the default Linux root:root.
+	# For /var/delphix, we only need to chown the directory itself
+	# since its contents should have been already chowned.
+	#
+	chown -R root: /var/dlpx-update /var/crash || errors=true
+	chown root: /var/delphix || errors=true
+
+	$errors && die "Some post-migration cleanup tasks have failed."
```

Comment:
out of curiosity, why don't we stop at the first error?

Reply:
The initial idea was for the cleanup to be "best-effort", and not to be a hard requirement of the upgrade, so we'd want to cleanup as much datasets as possible even if there was a failure. Now, since we have some mandatory cleanup tasks (like the `zfs promote` calls), we could either error out right away if anything fails, or only fail if a critical task fails. I'd probably opt for the former since it's simpler.